### PR TITLE
feat: add `sso` commands

### PR DESCRIPTION
**Pull request by hf**
Commands to manage Single Sign-On authentication on projects.

### CODE REVIEW

File: cmd/gen.go [hunk 1]
```diff
@@ -22,23 +22,23 @@ var (
...
 	}
 
-	local     bool
-	linked    bool
-	projectId string
-	dbUrl     string
-	schemas   []string
+	local        bool
+	linked       bool
+	genProjectId string
```
Comment (J0) [inline cmd/gen.go:27]: Are these changes related to SSO?

File: cmd/sso.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	ssoAttributeMappingFile string
+	ssoDomains              []string
+	ssoAddDomains           []string
+	ssoRemoveDomains        []string
+
+	ssoAddCmd = &cobra.Command{
+		Use:     "add <type> [flags]",
+		Short:   "Add a new SSO identity provider",
+		Args:    cobra.ExactArgs(1),
+		Example: `  supabase sso add saml --project-ref abcdefghijklmn --metadata-file ~/SAMLMetadata.xml`,
```
Comment (J0) [inline cmd/sso.go:26]: Would <type> here refer to the SSO method? (e.g. SAML, OIDC, password SSO etc) ?

File: cmd/sso.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return nil
+		},
+	}
+
+	ssoGetCmd = &cobra.Command{
+		Use:     "get <provider-id> [flags]",
+		Short:   "Get information about an SSO identity provider",
+		Args:    cobra.ExactArgs(1),
+		Example: `  supabase sso get b5ae62f9-ef1d-4f11-a02b-731c8bbb11e8 --project-ref abcdefghijklmn`,
```
Comment (J0) [inline cmd/sso.go:65]: What information would we be returning here? Perhaps can consider allowing a filter by specific attributes in the future since metadata files (and other xml files like attribute mapping) tend to be long and might flood the screen

File: cmd/sso.go (hunk 2 above)
Comment (hf) [reply cmd/sso.go:26]: Sorry what's "here" 🙃

File: cmd/sso.go (hunk 3 above)
Comment (hf) [reply cmd/sso.go:65]: All of it, probably we can also do something to fetch the metadata only, let's see.

File: cmd/sso.go (hunk 2 above)
Comment (J0) [reply cmd/sso.go:76]: Oops that should be `type`, but looks like it's been added with a default of saml.

File: internal/sso/update/update.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	"context"
+	"errors"
+	"os"
+
+	"github.com/supabase/cli/internal/sso/internal/render"
+	"github.com/supabase/cli/internal/utils"
+	"github.com/supabase/cli/pkg/api"
+)
+
+func Run(ctx context.Context, ref, providerId string, params api.UpdateProviderByIdJSONRequestBody, addDomains, removeDomains []string, format string) error {
```
Comment (sweatybridge) [inline internal/sso/update/update.go:13]: Let's validate the project ref here as a sanity check. for eg. https://github.com/supabase/cli/blob/main/internal/ssl_enforcement/get/get.go#L13

File: internal/sso/internal/render/render.go
Comment (sweatybridge) [inline internal/sso/internal/render/render.go:1]: Shall we name this as `internal/sso/render/render.go`? I think it's fine to have some internal packages not exposed as cmds.

File: cmd/status.go [hunk 5]
```diff
@@ -12,19 +12,9 @@ import (
...
-			status.OutputPretty,
-			status.OutputToml,
-			status.OutputYaml,
-		},
-		Value: status.OutputPretty,
-	}
-
+	override  []string
+	names     status.CustomName
+	output    = utils.OutputFlag(true /* allow env output */)
```
Comment (sweatybridge) [inline cmd/status.go:17]:
```suggestion
	override []string
	names    status.CustomName
	output   = utils.EnumFlag{
		Allowed: append([]string{status.OutputEnv, utils.OutputDefaultAllowed}),
		Value: status.OutputPretty,
	}
```

Can we do something like this ^? I think it reads more explicit than wrapping a function.

File: internal/utils/output.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+		return yaml.NewEncoder(w).Encode(value)
+
+	case OutputToml:
+		return toml.NewEncoder(w).Encode(value)
+
+	default:
+		return fmt.Errorf("Unsupported output encoding %q", format)
+	}
+
+	return nil
```
Comment (sweatybridge) [inline internal/utils/output.go:74]:
```suggestion
```

unnecessary return

File: cmd/sso.go [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+		PreRun: func(cmd *cobra.Command, args []string) {
+			cobra.CheckErr(cmd.MarkFlagRequired("project-ref"))
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var params api.UpdateProviderByIdJSONRequestBody
+
+			fsys := afero.NewOsFs()
+
+			if ssoMetadataFile != "" {
+				data, err := readMetadataFile(fsys, ssoMetadataFile)
```
Comment (sweatybridge) [inline cmd/sso.go:117]: Can we move file reading into internals? I think it's fine to build the full `api.UpdateProviderByIdJSONRequestBody` object inside internals too.

File: cmd/sso.go [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+				if err != nil {
+					return err
+				}
+
+				params.AttributeMapping = mapping
+			}
+
+			params.Domains = &ssoDomains
+
+			return update.Run(cmd.Context(), ssoProjectRef, args[0], params, ssoAddDomains, ssoRemoveDomains, ssoOutput.Value)
```
Comment (sweatybridge) [inline cmd/sso.go:139]: seems like `ssoAddDomains` and `ssoRemoveDomains` are unused by `update.Run`.

File: internal/sso/internal/render/render.go
Comment (hf) [reply internal/sso/internal/render/render.go:1]: Fine with me, just doesn't make sense to be usable outside sso.

File: internal/utils/output.go (hunk 6 above)
Comment (hf) [reply internal/utils/output.go:74]: AH THIS IS THE UNREACHABLE CODE! I kept looking for it since golangci doesn't tell you where it is!

File: cmd/sso.go (hunk 8 above)
Comment (hf) [reply cmd/sso.go:139]: Ah forgot about that, there should be a todo on it tho. Will use them.

File: cmd/sso.go [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+				if err != nil {
+					return err
+				}
+
+				params.MetadataXml = &data
+			} else if ssoMetadataURL != "" {
+				// TODO: fetch and validate Metadata
+				params.MetadataUrl = &ssoMetadataURL
+			} else {
+				return errors.New("--metadata-file or --metadata-xml must be provided")
```
Comment (kangmingtay) [inline cmd/sso.go:72]:
```suggestion
				return errors.New("--metadata-file or --metadata-url must be provided")
```

File: cmd/sso.go [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+
+			if ssoMetadataFile != "" {
+				data, err := readMetadataFile(fsys, ssoMetadataFile)
+				if err != nil {
+					return err
+				}
+
+				params.MetadataXml = &data
+			} else if ssoMetadataURL != "" {
+				// TODO: fetch and validate Metadata
```
Comment (kangmingtay) [inline cmd/sso.go:69]: hmm do we need to validate this or can we just leave it to gotrue?

File: cmd/sso.go (hunk 2 above)
Comment (kangmingtay) [reply cmd/sso.go:44]: i dont think this defaults to saml - if it's excluded, i get the following error: `Error: accepts 1 arg(s), received 0`

**Review by kangmingtay — CHANGES_REQUESTED** [review]

**Comment by hf** [comment]
@kangmingtay can't reply to that comment not sure why, so here it is:
> i dont think this defaults to saml - if it's excluded, i get the following error: Error: accepts 1 arg(s), received 0

This is intentional, the user has to enter `saml` or in the future `oidc` to explicitly specify the type.

File: cmd/sso.go (hunk 10 above)
Comment (hf) [reply cmd/sso.go:69]: Nothing too complex, but fetching the URL, checking it's returning 200 and XML should be enough. Also that it's HTTPS.

File: .gitignore [hunk 11]
```diff
@@ -4,6 +4,7 @@
 *.dll
 *.so
 *.dylib
+/cli
```
[the remark this answers is not in the document]
Comment (hf) [reply .gitignore:7]: Actually this should be `/cli` as that means ignore the `cli` entry at the root of the git repo.

File: .gitignore [hunk 12]
```diff
@@ -20,3 +21,4 @@
 # NPM
 node_modules
 package-lock.json
+
```
Comment (sweatybridge) [inline .gitignore:24]:
```suggestion
```

File: cmd/functions.go [hunk 13]
```diff
@@ -156,7 +158,7 @@ Enter your project ref: `, utils.GetSupabaseDashboardURL())
 
 		scanner := bufio.NewScanner(os.Stdin)
 		if !scanner.Scan() {
-			return errors.New("Cancelled " + utils.Aqua("supabase functions deploy") + ".")
+			return errors.New("Cancelled " + utils.Aqua(cmd.UseLine()) + ".")
```
Comment (sweatybridge) [inline cmd/functions.go:161]: We will probably refactor this separately as the Scan() may terminate due to either error and cancellation.

File: cmd/sso.go [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+		},
+	}
+
+	ssoRemoveCmd = &cobra.Command{
+		Use:     "remove <provider-id> [flags]",
+		Short:   "Remove an existing SSO identity provider",
+		Args:    cobra.ExactArgs(1),
+		Example: `  supabase sso remove b5ae62f9-ef1d-4f11-a02b-731c8bbb11e8 --project-ref mwjylndxudmiehsxhmmz`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return remove.Run(cmd.Context(), projectRef, args[0], ssoOutput.Value)
```
Comment (sweatybridge) [inline cmd/sso.go:88]: Is it possible to validate the providerId with regex?, ie. `args[0]`

File: cmd/sso.go [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+	ssoDomains              []string
+	ssoAddDomains           []string
+	ssoRemoveDomains        []string
+	ssoOutput               = utils.EnumFlag{
+		Allowed: utils.OutputDefaultAllowed,
+		Value:   utils.OutputPretty,
+	}
+
+	ssoAddCmd = &cobra.Command{
+		Use:     "add <type = saml> [flags]",
```
Comment (sweatybridge) [inline cmd/sso.go:60]: Might be a bit late to ask but can we make this an enum flag? ie. `sso add --type saml`

```go
	ssoType = utils.EnumFlag{
		Allowed: []string{"saml"},
		Value:   "saml",
	}
```

This would give us automatic validation and default value support.

Also, the `[flags]` part of UseLine will be added automatically by cobra

```suggestion
		Use:     "add",
```

File: cmd/sso.go [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+				MetadataURL:       ssoMetadataURL,
+				SkipURLValidation: ssoSkipURLValidation,
+				AttributeMapping:  ssoAttributeMappingFile,
+				Domains:           ssoDomains,
+			})
+		},
+	}
+
+	ssoRemoveCmd = &cobra.Command{
+		Use:     "remove <provider-id> [flags]",
```
Comment (sweatybridge) [inline cmd/sso.go:83]:
```suggestion
		Use:     "remove <provider-id>",
```

ditto

File: cmd/sso.go [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+		Short:   "Remove an existing SSO identity provider",
+		Args:    cobra.ExactArgs(1),
+		Example: `  supabase sso remove b5ae62f9-ef1d-4f11-a02b-731c8bbb11e8 --project-ref mwjylndxudmiehsxhmmz`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return remove.Run(cmd.Context(), projectRef, args[0], ssoOutput.Value)
+		},
+	}
+
+	ssoUpdateCmd = &cobra.Command{
+		Use:     "update <provider-id> [flags]",
```
Comment (sweatybridge) [inline cmd/sso.go:93]:
```suggestion
		Use:     "update <provider-id>",
```

File: cmd/sso.go [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+				AttributeMapping:  ssoAttributeMappingFile,
+				Domains:           ssoDomains,
+				AddDomains:        ssoAddDomains,
+				RemoveDomains:     ssoRemoveDomains,
+			})
+		},
+	}
+
+	ssoShowCmd = &cobra.Command{
+		Use:     "show <provider-id> [flags]",
```
Comment (sweatybridge) [inline cmd/sso.go:115]:
```suggestion
		Use:     "show <provider-id>",
```

File: internal/sso/get/get.go [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+	if resp.JSON200 == nil {
+		if resp.StatusCode() == http.StatusNotFound {
+			return fmt.Errorf("An identity provider with ID %q could not be found.", providerId)
+		}
+
+		return errors.New("Unexpected error fetching identity provider: " + string(resp.Body))
+	}
+
+	switch format {
+	case "metadata":
```
Comment (sweatybridge) [inline internal/sso/get/get.go:30]: Let's make a new constant for this special case in this file, ie. `const OutputMetadata = "metadata"`

**Review by sweatybridge — COMMENTED** [review]
Just a couple of comments.

File: cmd/sso.go (hunk 14 above)
Comment (kangmingtay) [reply cmd/sso.go:88]: yeah should be possible since it follows the uuid_v4 format

**Review by kangmingtay — APPROVED** [review]

File: cmd/sso.go (hunk 15 above)
Comment (hf) [reply cmd/sso.go:60]: Sure no problem. I was a bit skeptical of the positional argument myself.

**Review by sweatybridge — APPROVED** [review]
